pintc/src/predicate/optimize.rs

```diff
@@ -8,9 +8,9 @@ use crate::error::Handler;
 
 impl super::Contract {
     pub fn optimize(mut self, handler: &Handler) -> Self {
-        dead_code_elimination(handler, &mut self);
+        const_folding(&mut self); // TODO: Make passes independent. Right now we rely on const_folding before dead_constraint_elim
 
-        const_folding(&mut self);
+        dead_code_elimination(handler, &mut self);
 
         self
     }
```

Comment:
@mohammadfawaz in the past I believe you and I have talked about how optimization passes need to operate independently and not depend on the order they're run in.

I'm not sure if this violates that rule or not. To get the optimal final result, `const_folding` needs to be done before `dead_code_elimination`. Specifically for the case where the constraint is `constraint true || x` -> `constraint true` -> removed as dead code.

If we do it the other way around, we're left with the `constraint true` because we checked if the `constraint true || x` should be removed before it was transformed to `constraint true`.

The reason I think it may not violate the rule is because both passes can be run at anytime without causing the other to error. Technically they remain independent of each other, you just don't get the best final result unless they're paired together in the correct order.

The only way I can think of having the order no matter would be to duplicate the logic in `dead_code_elimination` that deals with `constraint true` and `constraint false` inside of `const_folding`. And at that point it seems like we're going to end up in a pattern of copying and pasting the various optimization passes everywhere.

Reply:
Yeah use the better ordering here... As we add more and more passes, you'll find that some passes need to be run multiple times in different places.

In practice, some orderings can be better than others. The rule we talked about emphasizes this statement:

> optimization passes need to operate independently and not depend on the order they're run in

This only means that each pass should be self contained and not dependent, **functionally**, on other passes. That is, you want the property that the passes _work_ (and do not crash etc.) regardless of which order they run in. That does not mean the produced code has to be exactly the same regardless of the order.